TAJO-1952: Implement PartitionFileFragment

blrunner opened the pull request:
This patch contains following modifications:

* Remove partition paths from PartitionedTableScanNode
* Implement PartitionedFileFragment
* Separate a method for pruning partition paths from PartitionedTableRewriter to PartitionedTableUtil
* Build a type by using partition name which contains partition keys and values

hyunsik commented on tajo-storage/tajo-storage-hdfs/src/main/java/org/apache/tajo/storage/FileTablespace.java:
```diff
@@ -477,12 +503,17 @@ public long getMinSplitSize() {
    * @throws IOException
    */
-  public List<Fragment> getSplits(String tableName, TableMeta meta, Schema schema, Path... inputs)
+  public List<Fragment> getSplits(String tableName, TableMeta meta, Schema schema, String[] partitions, Path... inputs)
```
I'm going to give a suggestion. We need to keep each method for one purpose if possible. That approach would keep logic simpler.

hyunsik commented:
Could you rebase it against the latest revision? I'd like to try the patch on my machine.

blrunner commented:
@hyunsik 

Thank you for your review. I've just rebased it against the latest version.

hyunsik commented on tajo-storage/tajo-storage-hdfs/src/main/proto/StorageFragmentProtos.proto:
```diff
@@ -32,3 +32,14 @@ message FileFragmentProto {
+  repeated string hosts = 5;
+  repeated int32 disk_ids = 6;
+  // Partition Name: country=KOREA/city=SEOUL
+  required string partitionName = 7;
```
In my opionion, partitionKeys would be more proper because the attribute includes concatenated partition keys.

hyunsik commented on tajo-storage/tajo-storage-hdfs/src/main/java/org/apache/tajo/storage/FileTablespace.java:
```diff
@@ -420,9 +421,29 @@ protected FileFragment makeSplit(String fragmentId, Path file, BlockLocation blo
     return new FileFragment(fragmentId, file, blockLocation);
   }
 
+  protected PartitionedFileFragment makePartitionSplit(String fragmentId, Path file, long start, long length,
```
It seems to be an unused method.

hyunsik commented on tajo-storage/tajo-storage-hdfs/src/main/java/org/apache/tajo/storage/fragment/PartitionedFileFragment.java:
```diff
@@ -0,0 +1,254 @@
+  @Expose public Long length; // required
+
+  private String[] hosts; // Datanode hostnames
+  @Expose private int[] diskIds;
```
Its getter and setter seem to be not used. Are there any reason to keep this information here?

blrunner commented:
@hyunsik 

Thank your for your review. I've just reflected your comments.

blrunner commented:
I updated the patch as following:

* Renamed PartitionedFileFragment to PartitionFileFragment
* Recovered codes for pruning partition in PartitionedTableRewrtier
* Recovered partition paths of PartitionedTableScanNode and added partition keys to PartitionedTableScanNode
* Implemented PartitionContent which contains partition paths, partition keys, partition volume

For the reference, I've tried to maintain existing codes as far as possible.

hyunsik commented on tajo-plan/src/main/java/org/apache/tajo/plan/partition/PartitionContent.java:
```diff
@@ -0,0 +1,69 @@
+  public PartitionContent() {
+  }
+
+  public PartitionContent(Path[] partitionPaths) {
```
unused ctor

hyunsik commented on tajo-plan/src/main/java/org/apache/tajo/plan/partition/PartitionContent.java:
```diff
@@ -0,0 +1,69 @@
+    this.partitionPaths = partitionPaths;
+  }
+
+  public PartitionContent(Path[] partitionPaths, long totalVolume) {
```
unused constructor

hyunsik commented on tajo-plan/src/main/java/org/apache/tajo/plan/rewrite/rules/PartitionedTableRewriter.java:
```diff
@@ -42,13 +45,13 @@
 import org.apache.tajo.storage.Tuple;
 import org.apache.tajo.storage.VTuple;
 import org.apache.tajo.util.StringUtils;
+import org.apache.tajo.util.TUtil;
```
unused

hyunsik commented on tajo-plan/src/main/java/org/apache/tajo/plan/rewrite/rules/PartitionedTableRewriter.java:
```diff
@@ -216,7 +228,22 @@ public String toString() {
+      totalVolume += summary.getLength();
+    }
+
+    partitionContent.setPartitionPaths(filteredPaths);
```
It can be refactored to use just the constructor instead of setters.

blrunner commented:
@hyunsik 

Updated the patch as following:

* Remove unused packages
* Remove PartitionedTableScanNode of PlanProto
* Remove unnecessary PartitionedTableScanNode usages of Repartitioner and PhysicalPlannerImpl

blrunner commented:
Added unit test cases for PartitionedTableRewriter

hyunsik commented on tajo-storage/tajo-storage-hdfs/src/main/java/org/apache/tajo/storage/FileTablespace.java:
```diff
@@ -550,6 +551,181 @@ public long getMinSplitSize() {
+    }
+
+    List<Map.Entry<String, Integer>> entries = new ArrayList<>(hostsBlockMap.entrySet());
+    Collections.sort(entries, new Comparator<Map.Entry<String, Integer>>() {
```
It can be simplified with lambda.

hyunsik commented on tajo-storage/tajo-storage-hdfs/src/main/java/org/apache/tajo/storage/FileTablespace.java:
```diff
@@ -550,6 +551,181 @@ public long getMinSplitSize() {
+  // The below code is for splitting partitioned table.
+  ////////////////////////////////////////////////////////////////////////////////
+
+  public PartitionFileFragment[] partitionSplit(String tableName, Path tablePath, String partitionKeys)
```
It seems to be not used.

hyunsik commented on tajo-storage/tajo-storage-hdfs/src/main/java/org/apache/tajo/storage/FileTablespace.java:
```diff
@@ -550,6 +551,181 @@ public long getMinSplitSize() {
+  protected Fragment makeNonPartitionSplit(String fragmentId, Path file, long start, long length,
+                                           BlockLocation[] blkLocations, String partitionKeys) throws IOException {
+
+    Map<String, Integer> hostsBlockMap = new HashMap<>();
```
This code fragment is duplicated to that of ``makeNonSplit``. Could you refactor it?

hyunsik commented on tajo-storage/tajo-storage-hdfs/src/main/java/org/apache/tajo/storage/FileTablespace.java:
```diff
@@ -550,6 +551,181 @@ public long getMinSplitSize() {
+        }
+      }
+      if(LOG.isDebugEnabled()){
+        LOG.debug("# of splits per partition: " + (splits.size() - previousSplitSize));
```
The log message is somewhat different from the actual meaning.

``PreviousSplitSize`` is the number of files in partition directories, and split is the number of total fragments. For example, there are two files in all partitions, and each file makes 5 fragments. Then, the total number of fragments is 10. In that case, 10 - 2 is jut 8. 8 is not the number of splits per partition.  The actual answer should be 5. In addition, we can say X per Y when X is the same across all Ys. It is hard to get answer because each partition file can be vary in terms of size. So, average is more proper here, I think.

hyunsik commented on tajo-storage/tajo-storage-hdfs/src/main/java/org/apache/tajo/storage/FileTablespace.java:
```diff
@@ -550,6 +551,181 @@ public long getMinSplitSize() {
+        files.addAll(listStatus(p));
+      }
+
+      int previousSplitSize = splits.size();
```
It would be better if the ```final``` marked for ``previousSplitSize``.

hyunsik commented on tajo-plan/src/main/java/org/apache/tajo/plan/partition/PartitionContent.java:
```diff
@@ -0,0 +1,45 @@
+import org.apache.hadoop.fs.Path;
+
+public class PartitionContent {
+  private Path[] partitionPaths;
```
It would be better if they all are marked by final.

blrunner commented:
@hyunsik 

Thank you for your detailed review. I reflected your comments.

blrunner commented:
@hyunsik 

I changed ```PartitionFileFragment``` by extending ```FileFragment``` because there are some codes for only ```FileFragment```. For example, you can see cast codes as following:
```
  public FileScanner(Configuration conf, final Schema schema, final TableMeta meta, final Fragment fragment) {
    this.conf = conf;
    this.meta = meta;
    this.schema = schema;
    this.fragment = (FileFragment)fragment;
    this.tableStats = new TableStats();
    this.columnNum = this.schema.size();
  }
```

blrunner commented:
Evaluated patch testing before and after ``PartitionFileFragment`` implementation as following:

* Dataset : TPCH-100G
* Tajo Cluster : 1 Master, 6 Workers
* Queries: Q1, Q3, Q5, Q6, Q7, Q8, Q9, Q10
* Tajo version for before  ``PartitionFileFragment`` implementation 
 -  0.11.1-SNAPSHOT
* Tajo version for after  ``PartitionFileFragment`` implementation:
 - 0.12.0-SNAPSHOT (partitions exist on catalog)
 - 0.12.0-SNAPSHOT (partitions doesn't exist on catalog)

The results were same as following:

* The number and order of execution blocks
* The number of stags in a each execution block
* The number of rows in a result
* All tuples in a result (excluded a few floating point value)

blrunner commented:
The Travis CI Build seems like a fail by another reason. Here is the pre-commit report by lasted patch as follows. https://builds.apache.org/job/PreCommit-TAJO-Build/900//console

blrunner commented:
@hyunsik 

Is there a problem that needs to be fixed?

blrunner commented:
@hyunsik 

I updated this PR as following:

* Remove the list of partition paths in PartitionedTableScanNode. It just contains the path of table root.
* PartitionedTableRewriter just rewrite ScanNode to PartitionedTableScanNode in BaseLogicalPlanRewriteRuleProvider.
* Actual partition pruning will be executed in Repartitioner by PartitionedTableRewriter.
